### PR TITLE
intercept ball

### PR DESCRIPTION
**Pull request by Snehal-Reddy**
Pull request bhul gaya

### CODE REVIEW

File: role/GoToBall.py [hunk 1]
```diff
@@ -139,14 +193,131 @@ def execute_fine_approach(self):
...
 
-
-
+    def intercept_complete(self):
+        ball_vel_angle = self.tan_inverse(self.kub.state.ballVel.y,self.kub.state.ballVel.x)
+        if ( abs(ball_vel_angle - self.tan_inverse(self.kub.state.ballPos.y-self.kub.state.homePos[self.kub.kubs_id].y , self.kub.state.ballPos.x-self.kub.state.homePos[self.kub.kubs_id].x)) < 0.0523599):
+            return 1
+        return 0
+
+    def getTime(self,pathLength):
```
Comment (saurabhkgp21) [inline role/GoToBall.py]: are you using this function in this file?

File: role/GoToBall.py [hunk 2]
```diff
@@ -62,6 +87,17 @@ def __init__(self,course_approch_thresh =  DISTANCE_THRESH/3,continuous=False):
 
         self.add_transition(GoToBall.State.fine_approach,
             behavior.Behavior.State.failed,lambda: self.behavior_failed,'failed')
+
+        self.add_transition(GoToBall.State.intercept,
+            behavior.Behavior.State.failed,lambda: self.behavior_failed,'failed')
+
+    def tan_inverse(self,y,x):
```
Comment (saurabhkgp21) [inline role/GoToBall.py]: @Snehal-Reddy I think it will be better to add this function into ```utils/math_function.py```

File: role/GoToBall.py [hunk 3]
```diff
@@ -39,21 +43,42 @@ def __init__(self,course_approch_thresh =  DISTANCE_THRESH/3,continuous=False):
...
 
         self.add_transition(GoToBall.State.setup,
-            GoToBall.State.fine_approach,lambda: self.fine_approach(),'ball_in_vicinity')
+           GoToBall.State.intercept,lambda:self.ball_moving(),'intercept_ball')
+
+        self.add_transition(GoToBall.State.setup,
+            GoToBall.State.fine_approach,lambda: self.fine_approach() and not self.ball_moving() ,'ball_in_vicinity')
+
+        self.add_transition(GoToBall.State.setup,
+            GoToBall.State.course_approach,lambda: self.course_approach() and not self.ball_moving(),'setup')
```
Comment (saurabhkgp21) [inline role/GoToBall.py:60]: @Snehal-Reddy will this work when ball will be moving in real play?

File: role/GoToBall.py [hunk 4]
```diff
@@ -39,21 +43,42 @@ def __init__(self,course_approch_thresh =  DISTANCE_THRESH/3,continuous=False):
...
 
         self.add_transition(GoToBall.State.setup,
-            GoToBall.State.course_approach,lambda: self.course_approach(),'setup')
+            GoToBall.State.intercept,lambda:self.ball_moving(),'intercept_ball')
+
+        self.add_transition(GoToBall.State.intercept,
+            GoToBall.State.course_approach,lambda:self.course_approach() and not self.ball_moving(),'setup')
```
Comment (saurabhkgp21) [inline role/GoToBall.py:66]: @Snehal-Reddy .

File: role/GoToBall.py (hunk 1 above)
Comment (Snehal-Reddy) [reply role/GoToBall.py]: No. I will remove it.

File: role/GoToBall.py (hunk 2 above)
Comment (Snehal-Reddy) [reply role/GoToBall.py]: Okay done

File: role/GoToBall.py (hunk 3 above)
Comment (Snehal-Reddy) [reply role/GoToBall.py:60]: It was working

**Comment by saurabhkgp21** [comment]
Closes #23

**Pull request merged**